Add logging infrastructure

antkmsft opened the pull request.

JeffreyRichter commented on sdk/core/core/inc/az_log.h:
```diff
@@ -0,0 +1,48 @@
+typedef struct {
+  clock_t slow_response_threshold; // How long should it take the HTTP request to start being
+                                   // considered a slow response (AZ_LOG_SLOW_RESPONSE)
+} az_log_options;
```
We'll need a .c file for the log pipeline policy. When we have this file, this az_log_options struct belongs in that file; not in this file. Also, we do not want to use clock_t because we may not want to take a dependency on it and because it is in ticks; I think we should use a 32-bit int instead which represents milliseconds.

antkmsft replied:
Done.

antkmsft replied:
Milliseconds are done.

antkmsft replied:
Moved the slow response threshold to .c file.
Done your suggestion in full.

JeffreyRichter commented on sdk/core/core/inc/az_log.h:
```diff
@@ -0,0 +1,48 @@
+                                   // considered a slow response (AZ_LOG_SLOW_RESPONSE)
+} az_log_options;
+
+typedef void (*az_log)(az_log_classification const classification, char const * const message);
```
This is a public typedef

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/inc/az_log.h:
```diff
@@ -0,0 +1,48 @@
+
+typedef void (*az_log)(az_log_classification const classification, char const * const message);
+
+void az_log_set_classifications(
```
This is a public function

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/inc/az_log.h:
```diff
@@ -0,0 +1,48 @@
+    az_log_classification const * const classifications,
+    size_t const classifications_length);
+
+void az_log_reset_classifications();
```
This is a private function and we may not need it at all; it was for testing in Go and we may not need it for C. I'd remove it for now and only add it back if we really need it. I'm not sure that Go actually needs it either.

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/inc/az_log.h:
```diff
@@ -0,0 +1,48 @@
+
+void az_log_reset_classifications();
+
+void az_log_set_listener(az_log * const listener);
```
This is a public function

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/inc/az_log.h:
```diff
@@ -0,0 +1,48 @@
+
+void az_log_set_listener(az_log * const listener);
+
+void az_log_write(az_log_classification const classification, char const * const message);
```
This is a friend/internal function so it needs to be in a different header file

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/inc/az_log.h:
```diff
@@ -0,0 +1,48 @@
+
+void az_log_write(az_log_classification const classification, char const * const message);
+
+bool az_log_should_write(az_log_classification const classification);
```
This is a friend/internal function so it needs to be in a different header file

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -5,12 +5,16 @@
 #include <az_url.h>
 
+#include <stddef.h>
+#include <time.h>
```
You may not need this if we get rid of clock_t

antkmsft replied:
I think we are going to need it in the implementation - even though users will be specifying threshold in milliseconds, the time we get from `clock()` is in `clock_t` and there is `CLOCKS_PER_SEC` for me to do the conversions necessary.

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/src/az_log.c:
```diff
@@ -0,0 +1,44 @@
+  _az_log_classifications_length = classifications_length;
+}
+
+void az_log_reset_classifications() {
```
Comment out for now.

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/src/az_log.c:
```diff
@@ -0,0 +1,44 @@
+
+void az_log_set_listener(az_log * const listener) { _az_log_listener = listener; }
+
+void az_log_write(az_log_classification const classification, char const * const message) {
```
IMPORTANT: We'll probably need to add thread safety to ALL these functions.

JeffreyRichter replied:
Yes, the char* should be a span

antkmsft replied:
Done char => span

antkmsft replied:
I put the TODO comments, and created https://github.com/Azure/azure-sdk-for-c/issues/316
I consider I've done your suggestion in scope of this PR.

JeffreyRichter commented on sdk/core/core/src/az_log.c:
```diff
@@ -0,0 +1,44 @@
+}
+
+bool az_log_should_write(az_log_classification const classification) {
+  if (_az_log_classifications == NULL || _az_log_classifications_length == 0) {
```
Please add a comment about the intent here: "If the user hasn't registered any classifications, then we log everything."
Now that I think about it, we may not want this behavior. What should default logging be? Maybe nothing. I'll talk to the IoT team about this.

antkmsft replied:
I added the comment. Maybe it is fine to log everything by default, as opposed to nothing.
1. In case user wants no log, they don't set the listener.
2. In case user wants everything, they don't specify the classifications. So, when we start logging new types of classifications, user starts to get them as well.

Ideally, for the maximum flexibility, there should be a way to include and exclude classifications, i.e. "I want everything except the slow response errors". And then when we add new types of classifications, user starts getting them as well. Or, if user told us they want only the retry and error messages, they get only these types of log entries all the time, and nothing more. Or, user may want to only get the keyvault errors, and nothing else.

For now, I think, our approach is fine, until we know of better scenarios.
But still, did you have a chance to talk to the IoT team about this?

JeffreyRichter commented on sdk/core/core/src/az_log.c:
```diff
@@ -0,0 +1,44 @@
+    return true;
+  }
+
+  for (size_t i = 0; i < _az_log_classifications_length; ++i) {
```
Please add comment about intent

antkmsft replied:
Done.

JeffreyRichter commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
   return AZ_OK;
 }
 
+AZ_INLINE uint8_t _az_decimal_to_ascii(uint8_t const d) {
```
I don't think we need this; I think you can pass a uint8_t to a function that wants a long long

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+  return '0' + d;
+}
+
+AZ_INLINE AZ_NODISCARD az_result
```
```suggestion
static AZ_NODISCARD az_result
```

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+}
+
+AZ_INLINE AZ_NODISCARD az_result
+_az_span_builder_append_unsigned_number(az_span_builder * const self, unsigned long long const n) {
```
```suggestion
_az_span_builder_append_unsigned_number(az_span_builder * const self, uint64_t const n) {
```

antkmsft replied:
The intention here is to handle `unsigned long long`, the biggest integer type you can get using the standard syntax, regardless of the actual "bitness" - the code will handle any (well, 3 bits minimum)

antkmsft replied:
Done

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+    return az_span_builder_append_byte(self, '0');
+  }
+
+  static unsigned long long div_max = 1;
```
```suggestion
  static uint64_t div_max = 1;
```

antkmsft replied:
Should we end up using `int*_t`, we then know `div_max` right away, and no need to calculate it. But think we should use `unsigned long long`, and not `uint64_t`.

sergey-shandar replied:
Most of the time, when we write/read data, we would like to be as much deterministic as possible. There are only few platform-dependant things that we use and, usually, they temporary and not serializable. For example, pointers, `size_t` (a size of array in memory), `ptrdiff_t` (a difference between pointers). Everything else most-likely should have constant size on all platforms.

antkmsft replied:
Done.

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+
+  static unsigned long long div_max = 1;
+  if (div_max == 1) {
+    for (unsigned long long next = 10; next > div_max; next *= 10) {
```
```suggestion
    for (uint64_t next = 10; next > div_max; next *= 10) {
```

antkmsft replied:
Done

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+    }
+  }
+
+  unsigned long long div = div_max;
```
```suggestion
uint64_t div = div_max;
```

antkmsft replied:
Done.

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+  }
+
+  unsigned long long div = div_max;
+  unsigned long long nn = n;
```
```suggestion
  uint64_t nn = n;
```

antkmsft replied:
Done.

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+}
+
+AZ_NODISCARD az_result
+az_span_builder_append_unsigned_number(az_span_builder * const self, unsigned long long const n) {
```
```suggestion
az_span_builder_append_unsigned_number(az_span_builder * const self, uint64_t const n) {
```

antkmsft replied:
Done.

sergey-shandar commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
+}
+
+AZ_NODISCARD az_result
+az_span_builder_append_signed_number(az_span_builder * const self, long long const n) {
```
```suggestion
az_span_builder_append_signed_number(az_span_builder * const self, int64_t const n) {
```

antkmsft replied:
Done.

vhvb1989 commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -92,8 +96,35 @@ AZ_NODISCARD az_result az_http_pipeline_policy_logging(
+  }
+
+  clock_t const threshold
+      = (data == NULL) ? 0 : ((az_log_options const *)data)->slow_response_threshold;
```
data can't never be == NULL here.  when it is null, we are returning next policy in the line above

antkmsft replied:
You're right, I rewrote that part now.

vhvb1989 commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -92,8 +96,35 @@ AZ_NODISCARD az_result az_http_pipeline_policy_logging(
+
+  clock_t const end = clock();
+
+  if (data != NULL) {
```
same thing here. Am I missing something?  I wouldn't expect data to be NULL at this point (due to line 99)

antkmsft replied:
You're right, I rewrote that part now.

vhvb1989 commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -92,8 +96,35 @@ AZ_NODISCARD az_result az_http_pipeline_policy_logging(
+
+  clock_t const end = clock();
+
+  if (data != NULL) {
```
same thing here. Am I missing something?  I wouldn't expect data to be NULL at this point (due to line 99

antkmsft replied:
You're right, I rewrote that part now.

vhvb1989 commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -92,8 +96,35 @@ AZ_NODISCARD az_result az_http_pipeline_policy_logging(
+
+  clock_t const end = clock();
+
+  if (data != NULL) {
```
same thing here. Am I missing something?  I wouldn't expect data to be NULL at this point (due to line 99

antkmsft replied:
You're right, I rewrote that part now.

vhvb1989 commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -92,8 +96,35 @@ AZ_NODISCARD az_result az_http_pipeline_policy_logging(
+
+      AZ_RETURN_IF_FAILED(az_span_builder_append(&log_msg_bldr, AZ_STR(" ms\0")));
+
+      az_log_write(AZ_LOG_SLOW_RESPONSE, (char const *)az_span_builder_result(&log_msg_bldr).begin);
```
wondering..  can we have logging api to work with spans instead of char ?   
It might be less work to delegate converting span to 0-terminated str to logging, instead of taking care of this conversion from everywhere where we want to log something from one span builder.

Otherwise, why can't we just use a char array here directly instead of constructing the span?

antkmsft replied:
Done.

vhvb1989 commented on sdk/core/core/src/az_span_builder.c:
```diff
@@ -23,6 +25,59 @@ AZ_NODISCARD az_result az_span_builder_append_byte(az_span_builder * const self,
 }
 
+AZ_INLINE uint8_t _az_decimal_to_ascii(uint8_t const d) {
+  assert(d < 10);
```
wondering...  why assert would be better than our az_contract here?

antkmsft replied:
it is a single line function that is only called a few lines below, assert shoud never be triggered because we should do mod 10. It is only if things go super wrong and whoever modifies it (why?) basically not paying attention at all.
AZ_CONTRACT seems a bit overkill, and to me it more looks like its purpose is to check the externally passed arguments whose values are beyond our control;

JeffreyRichter commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -5,12 +5,16 @@
+
 #include <_az_cfg.h>
 
 AZ_NODISCARD AZ_INLINE az_result az_http_pipeline_nextpolicy(
```
This code isn't right. The logic should be like this:
- if should log (Request), create log entry with HTTP method, URL/path, query parameters and headers and then log it.
- Get current time and save it
- Call next policy
- Get current time and subtract previous time to get operation duration
- if should log (slow response) && duration > slowResponseThreshold, then log with HTTP response status line & response headers. Also, include the original HTTP method, url/path, QPs, and request headers
- if should log (Response), then log the same entry again (don't construct it twice though to save CPU).
- return from this policy

antkmsft replied:
Done, please check out the new code in this file, and in the az_log.c

JeffreyRichter commented on sdk/core/core/src/az_http_policy.c:
```diff
@@ -92,8 +96,35 @@ AZ_NODISCARD az_result az_http_pipeline_policy_logging(
+      az_span_builder log_msg_bldr
+          = az_span_builder_create((az_mut_span)AZ_SPAN_FROM_ARRAY(log_msg_buf));
+
+      AZ_RETURN_IF_FAILED(az_span_builder_append_unsigned_number(
```
Failure to log MUST not fail the operation. If the buffer is not large enough to hold the log entry, just log what you have and keep going.

antkmsft replied:
Done.

antkmsft commented on sdk/core/core/test/test_log.c:
```diff
@@ -0,0 +1,245 @@
+
+  {
+    _reset_log_invocation_status();
+    az_log_set_listener(&_log_listener);
```
@JeffreyRichter, do you like how the usage code in this file looks like?
If you think it is good to go, I have nothing more to add to this PR. I addressed all the prior comments.

antkmsft commented:
@JeffreyRichter, I reacted to all feedback, most importantly I removed the slow response, and added `az_clock_internal.h`. I think I this is ready to go.
